### PR TITLE
[libcxx] Align `__recommend() + 1`  by __endian_factor

### PR DESCRIPTION
**Pull request by vitalybuka**
This is detected by asan after #83774

Allocation size will be divided by `__endian_factor` before storing. If it's not aligned,
we will not be able to recover allocation size to pass into  `__alloc_traits::deallocate`.

we have code like this 
```
 auto __allocation = std::__allocate_at_least(__alloc(), __recommend(__sz) + 1);
    __p               = __allocation.ptr;
    __set_long_cap(__allocation.count);

void __set_long_cap(size_type __s) _NOEXCEPT {
    __r_.first().__l.__cap_     = __s / __endian_factor;
    __r_.first().__l.__is_long_ = true;
  }

size_type __get_long_cap() const _NOEXCEPT {
    return __r_.first().__l.__cap_ * __endian_factor;
  }

inline ~basic_string() {
    __annotate_delete();
    if (__is_long())
      __alloc_traits::deallocate(__alloc(), __get_long_pointer(), __get_long_cap());
  }
```
1. __recommend() -> even size
2. `std::__allocate_at_least(__alloc(), __recommend(__sz) + 1)` - > not even size
3. ` __set_long_cap() `- > lose one bit of size for __endian_factor == 2 (see `/ __endian_factor`)
4. `__alloc_traits::deallocate(__alloc(), __get_long_pointer(), __get_long_cap())` -> uses even size (see `__get_long_cap`)

### CODE REVIEW

**Review by philnik777 — CHANGES_REQUESTED** [review]
Could you add some more info here? I don't see what bug this fixes, and I don't understand why we can't drop a byte when setting the long cap.

**Comment by jyknight** [comment]
We pass the capacity directly to deallocate, e.g.:
```
  inline _LIBCPP_CONSTEXPR_SINCE_CXX20 ~basic_string() {
    __annotate_delete();
    if (__is_long())
      __alloc_traits::deallocate(__alloc(), __get_long_pointer(), __get_long_cap());
  }
```
and for sized-delete, we are required to pass the _actual_ size that was passed to allocate, not a rounded amount.

**Comment by philnik777** [comment]
AFAICT we always return `__align_it() - 1` from `__recommend` though? Adding 1 would bring us back to an aligned size. (Don't ask me why it's written this way)

**Comment by vitalybuka** [comment]
> AFAICT we always return `__align_it() - 1` from `__recommend` though? Adding 1 would bring us back to an aligned size. (Don't ask me why it's written this way)

The assert inserted ther confirm that it's not true in some cases, on check-cxx, even without sanitizers.

**Comment by philnik777** [comment]
> > AFAICT we always return `__align_it() - 1` from `__recommend` though? Adding 1 would bring us back to an aligned size. (Don't ask me why it's written this way)
> 
> The assert inserted there confirms that it's not true in some cases, on check-cxx, even without sanitizers.

Ok, then what am I missing?

**Comment by philnik777** [comment]
BTW it's expected that there might be a byte dropped if the allocator returns more than what we asked for. That's completely legal for us to do. (But I don't think we test that for `string`, so it shouldn't affect the tests)

**Comment by vitalybuka** [comment]
> > > AFAICT we always return `__align_it() - 1` from `__recommend` though? Adding 1 would bring us back to an aligned size. (Don't ask me why it's written this way)
> > 
> > 
> > The assert inserted there confirms that it's not true in some cases, on check-cxx, even without sanitizers.
> 
> Ok, then what am I missing?

Sorry, maybe we are discussing different things.
Are you questioning particular case? Some __align_it<> from this patch may be unnecessary.
but most stuff like this:

```
 auto __allocation = std::__allocate_at_least(__alloc(), __recommend(__sz) + 1);
    __p               = __allocation.ptr;
    __begin_lifetime(__p, __allocation.count);
    __set_long_pointer(__p);
    __set_long_cap(__allocation.count);
    __set_long_size(__sz);
```

1.  __recommend() -> even size
2. std::__allocate_at_least(__alloc(), __recommend(__sz) + 1) - > not even size
3. __set_long_cap() - > lose one bit of size for __endian_factor == 2
4. __alloc_traits::deallocate(__alloc(), __get_long_pointer(), __get_long_cap()); -> uses even size

**Comment by vitalybuka** [comment]
> BTW it's expected that there might be a byte dropped if the allocator returns more than what we asked for. That's completely legal for us to do. (But I don't think we test that for `string`, so it shouldn't affect the tests)

Hm, I would expect allocator should still accept the same size for new/delete, even if it returns more. Otherwise using those operators will be very annoying.

**Comment by philnik777** [comment]
> > BTW it's expected that there might be a byte dropped if the allocator returns more than what we asked for. That's completely legal for us to do. (But I don't think we test that for `string`, so it shouldn't affect the tests)
> 
> Hm, I would expect allocator should still accept the same size for new/delete, even if it returns more. Otherwise using those operators will be very annoying.

Yes, it does. With `allocate_at_least` you can give it anything between the requested size and the actually allocated size. Because of that it's fine for us to drop a byte that's been promised by the allocator.

**Comment by philnik777** [comment]
I'm asking where we get the mismatch. My current mental model is that we always allocate an even number of elements (due to `__recommend()` returning an odd number plus one resulting in an even number), which apparently isn't the case. In which part of the code is my mental model failing? Your current solution looks like a band-aid to me and not like it's fixing the underlying issue.

**Comment by vitalybuka** [comment]
> I'm asking where we get the mismatch. My current mental model is that we always allocate an even number of elements (due to `__recommend()` returning an odd number plus one resulting in an even number), which apparently isn't the case. In which part of the code is my mental model failing? 

Got it, will reproduce with details soon.

> Your current solution looks like a band-aid to me and not like it's fixing the underlying issue.

Sorry, it's just how I understood the code, will be happy to improve.

File: libcxx/include/string
Comment (vitalybuka) [inline libcxx/include/string:1879]: @philnik777 
Note, this assert is enough to reproduce the issue on `LIT_FILTER=string ninja check-cxx`, in case it would be easy for you do debug yourself.

Comment (vitalybuka) [reply libcxx/include/string:1879]: Or maybe not...
I'll be back with details soon.

**Comment by philnik777** [comment]
I've looked into this a bit more and it turns out that we have two unrelated bugs in `__recommend`. First, we add one if `__guess == __min_cap`, and second our minimum alignment we give to `__align_it` is one, instead of the `__endian_factor`.

I think
```diff
diff --git a/libcxx/include/string b/libcxx/include/string
index 883bc1d7e5dc..13af92fbce3c 100644
--- a/libcxx/include/string
+++ b/libcxx/include/string
@@ -1959,10 +1959,10 @@ private:
     if (__s < __min_cap) {
       return static_cast<size_type>(__min_cap) - 1;
     }
-    const size_type __boundary = sizeof(value_type) < __alignment ? __alignment / sizeof(value_type) : 1;
+    const size_type __boundary = sizeof(value_type) < __alignment ? __alignment / sizeof(value_type) : __endian_factor;
     size_type __guess          = __align_it<__boundary>(__s + 1) - 1;
     if (__guess == __min_cap)
-      ++__guess;
+      __guess += 2;
     return __guess;
   }
```
should fix the issues you're seeing.

File: libcxx/include/string
Comment (vitalybuka) [reply libcxx/include/string:1879]: I see:

https://github.com/llvm/llvm-project/actions/runs/8899718744/job/24441331959?pr=90645

Also my local run if we apply only #90645
```
gdb runtimes/runtimes-bins/test/std/strings/basic.string/string.modifiers/string_replace/Output/size_size_pointer_size.pass.cpp.dir/t.tmp.exe
```

```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x00007ffff7b691cf in __pthread_kill_internal (signo=<optimized out>, threadid=<optimized out>) at ./nptl/pthread_kill.c:89
#2  __GI___pthread_kill (threadid=<optimized out>, signo=<optimized out>) at ./nptl/pthread_kill.c:89
#3  0x00007ffff7cb3e90 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#4  0x00007ffff7b054b2 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff7f44bce in std::__1::__libcpp_verbose_abort (format=0x7ffff7e9c1f8 "%s")
    at /usr/local/google/home/vitalybuka/src/llvm.git/llvm-project/libcxx/src/verbose_abort.cpp:74
#6  0x00007ffff7f2da14 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__set_long_cap[abi:de190000](unsigned long) (this=0x7fffffffc7b0, __s=25) at include/c++/v1/string:1879
#7  0x00007ffff7f3ab96 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__init (this=0x7fffffffc7b0, 
    __s=0x5555555e638d "1234567890123456789bcde", __sz=23) at include/c++/v1/string:2233
#8  0x0000555555584334 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::basic_string[abi:de190000]<0>
    ()
#9  0x0000555555557441 in bool test0<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >() ()
#10 0x00005555555552d9 in void test<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >() ()
#11 0x00005555555552bb in main ()
```

1. __init(const value_type* __s, size_type __sz = 23)
2. __recommend(23) returns 24, because 23 is `__min_cap`
3. then as in the description we allocate 25, but store  `25 / 2`
4. and next time we delete 24

**Comment by vitalybuka** [comment]
> I've looked into this a bit more and it turns out that we have two unrelated bugs in `__recommend`. First, we add one if `__guess == __min_cap`, and second our minimum alignment we give to `__align_it` is one, instead of the `__endian_factor`.
> 
> I think
> 
> ```diff
> diff --git a/libcxx/include/string b/libcxx/include/string
> index 883bc1d7e5dc..13af92fbce3c 100644
> --- a/libcxx/include/string
> +++ b/libcxx/include/string
> @@ -1959,10 +1959,10 @@ private:
>      if (__s < __min_cap) {
>        return static_cast<size_type>(__min_cap) - 1;
>      }
> -    const size_type __boundary = sizeof(value_type) < __alignment ? __alignment / sizeof(value_type) : 1;
> +    const size_type __boundary = sizeof(value_type) < __alignment ? __alignment / sizeof(value_type) : __endian_factor;
>      size_type __guess          = __align_it<__boundary>(__s + 1) - 1;
>      if (__guess == __min_cap)
> -      ++__guess;
> +      __guess += 2;
>      return __guess;
>    }
> ```
> 
> should fix the issues you're seeing.

My last reply written before reading this, but looks like they are consistent.

File: libcxx/include/string
Comment (vitalybuka) [inline libcxx/include/string:1966]: works with ASSERT, and likely will work with asan
Should `__guess += 2` be replaced with  `__guess += __endian_factor` if we want to keep  __endian_factor alignment?

Comment (vitalybuka) [reply libcxx/include/string:1966]: > works with ASSERT, and likely will work with asan Should `__guess += 2` be replaced with `__guess += __endian_factor` if we want to keep __endian_factor alignment?

Assuming __endian_factor will be `>2` is unlikely, 2 will work, but __endian_factor is better for reader?

Comment (philnik777) [reply libcxx/include/string:1966]: Yeah, I think `__endian_factor` would be the right call. Maybe we should rename it into something like `__min_capacity_alignment`, since that's what it basically is? Just a though, not a request for this patch.

**Review by philnik777 — COMMENTED** [review]
To unblock landing #83774 I'm happy with this patch if you remove the `_LIBCPP_ASSERT_INTERNAL` (since it may not actually be correct) and change `__guess += 2` to `__guess += __endian_factor`.

File: libcxx/include/string
Comment (vitalybuka) [reply libcxx/include/string]: Done

**Comment by vitalybuka** [comment]
> _LIBCPP_ASSERT_INTERNAL

Done, but I am curious why this can be wrong?
Seems set/get should return exactly the same value.

**Comment by philnik777** [comment]
> > _LIBCPP_ASSERT_INTERNAL
> 
> Done, but I am curious why this can be wrong? Seems set/get should return exactly the same value.

As I tried to explain before, `__allocate_at_least` may return more bytes than we requested with a custom allocator - possibly (even though unlikely) an odd number of elements. Since we drop the last bit of our capacity, we have no choice but to drop the last element of that allocation.

**Review by philnik777 — APPROVED** [review]
LGTM with green CI.

**Comment by philnik777** [comment]
Oh, and please update the commit message.

**Comment by vitalybuka** [comment]
> > > _LIBCPP_ASSERT_INTERNAL
> > 
> > 
> > Done, but I am curious why this can be wrong? Seems set/get should return exactly the same value.
> 
> As I tried to explain before, `__allocate_at_least` may return more bytes than we requested with a custom allocator - possibly (even though unlikely) an odd number of elements. Since we drop the last bit of our capacity, we have no choice but to drop the last element of that allocation.

Hm, then as-is custom allocator can return more, but if allocator needs correct value for __alloc_traits::deallocate, `allocate` suppose to return size magically aligned.

**Comment by EricWF** [comment]
Could you please add tests for this change?

**Comment by vitalybuka** [comment]
> Could you please add tests for this change?

There are plenty of tests that cover this case. it's just invisible on regular builds, but it will fail with sanitizers.
I tried to add assert, but according to https://github.com/llvm/llvm-project/pull/90292#issuecomment-2089669082 it's invalid.

**Comment by vitalybuka** [comment]
> > Could you please add tests for this change?
> 
> There are plenty of tests that cover this case. it's just invisible on regular builds, but they will fail with asan. I tried to add assert, but according to [#90292 (comment)](https://github.com/llvm/llvm-project/pull/90292#issuecomment-2089669082) it's invalid.

I will check if it's testable using __alloc_traits.

**Comment by MBkkt** [comment]
@vitalybuka 

Hello, definitely this patch fix the issue, but I have few questions
1. Why do we need to allocate 25/26 bytes for std::string (abi version 1, big endian/little endian) of size 23? Instead of 24
2. Why do we need assign  (allocation_size / 2 * 2) to capacity instead of just assign capacity? I know it's kind of fixed for new abi and little endian, but big endian code still not good

**Comment by MBkkt** [comment]
@vitalybuka 

https://github.com/llvm/llvm-project/pull/92652/files

**Comment by vitalybuka** [comment]
> Hello, definitely this patch fix the issue, but I have few questions

Probably @philnik777 is better for answering this question.

**Comment by alk** [comment]
Hi. I came here via https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=279560. This fix is not "just" about unbreaking some asan flag flip. It fixes real bug.

I.e. sized deallocation is intended to be a performance feature where memory allocator is given original allocation size. For ~most modern allocators (e.g. both versions of tcmalloc, jemalloc, mimalloc and perhaps more) this lets them obtain allocation's size-class directly from size rather than looking it up in some metadata. It saves cycles. With libc++ lying about the size, it is inviting the difficult to diagnose crashing bug.

I see that libc++-18 as shipped by Debian has this bug. libc++-17 and earlier appears to be fine (although I only did fairly simple test: https://paste.debian.net/hidden/8edbffc1/).

From what I see,  https://github.com/llvm/llvm-project/commits/release/18.x/ is your llvm/libc++ 18 release branch and I don't see this fix backported there. Can someone make sure it is cherry-picked?

BTW do I guess correctly that Google's internal codebase (which definitely runs on -fsized-deallocation) didn't catch this because Google runs on alternative string abi layout thingy? Otherwise, I'd be worried how come this wasn't caught much sooner.

**Comment by philnik777** [comment]
Thanks for the info. Though I'd be surprised if this wasn't a bug in 17. The code this patch touches hasn't been meaningfully changed in over six years. @tstellar is another release planned?

**Comment by alk** [comment]
Gentle ping. As people around the world get their libc++ upgraded to 18 they'll be getting broken library. We really need this fix backported to 18.x.

CC-ing @ckennelly for awareness (and some llvm connections to ping). It affects all C++ allocators offering sized delete support including Google's.

**Comment by ldionne** [comment]
I don't think another release is planned. If this is truly a regression from 17, we could ask @tstellar for another release if this is serious enough.

@alk It would be helpful if you can confirm whether that is a regression from 17. I pasted your reproducer in Godbolt but didn't pay close attention to what it does. From the looks of it, Clang 16-17-18 all behave the same: https://godbolt.org/z/oq3v6Yvqa

**Comment by alk** [comment]
Yes, it is regression from 17 and all previous releases. Again, this has nothing to do with asan flag flip. Bug was introduced that'll catch anyone who do -fsized-deallocation and with C++ allocators that don't ignore size, and this commit fixes it.

**Comment by ldionne** [comment]
@alk Sorry, I am failing to see *how* this is a regression introduced in 18. Can you share a godbolt link, commit or something else?

**Comment by alk** [comment]
"How" is up to you guys. I know that simple test posted above works on libc++ 17 and fails on libc++ 18. Shouldn't it be enough? Here is link again: https://paste.debian.net/hidden/8edbffc1/

Some kind of crash will happen on any C++ memory allocator that cares to verify sizes. gperftools on master does in debug builds. "Abseil" tcmalloc from github.com/google/tcmalloc has been doing this for ages. So you should be able to verify regression in multiple ways.

At FreeBSD ticket Dimitry Andric claims that he found the commit(s) (kudos and much thanks for this; @DimitryAndric). Here: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=279560#c10 But it isn't super trivial to me, so I'll leave it to experts.

"how" might be worth mini-retrospective, especially given that I know for sure that Google does sized delete internally on 100% internal workloads.

**Comment by DimitryAndric** [comment]
As far as I could determine, this particular behavior of std::string regressed with 04ce0baf015c ("Unconditionally lower std::string's alignment requirement from 16 to 8"), which also got merged to 18.x, but it could be that it does not manifest itself on all platforms. We did observe the issue on FreeBSD, which uses libc++ by default, and @alk observed it on Debian Linux.

**Comment by ldionne** [comment]
@alk Ah, I managed to make your reproducer work. I was using incorrect flags, it needs optimizations enabled to trigger. This is a slightly simplified version that shows the issue on Clang 18 but not on 17 and 16: https://godbolt.org/z/nYYvG74d7

@DimitryAndric Thanks a lot for the information.

So then this is indeed a regression in LLVM 18. Since this is happening in `std::string` which is very widely used, I believe this may have enough impact to be worth another 18.1.x release. WDYT?

**Comment by tstellar** [comment]
There are now more 18.1.x releases planned, but we can do another one if the bug is severe enough.  Can someone summarize for me what kinds of use cases are impacted?

**Comment by ldionne** [comment]
Basically, folks passing the `-fsized-deallocation` Clang flag (which is off by default in LLVM 18) will get the wrong deallocation size passed to their `operator delete` when storage has been allocated/deallocated by `std::string`. I don't know what the result of passing the wrong size will usually be, I guess it depends on what the custom `operator delete` does, but I could definitely imagine it being a correctness concern.

@alk @ckennelly There seems to be a claim that Google's using this flag at large scale internally. If it's possible to detail the impact of this regression a bit more, it might help to give at least one "real world" point of view on this.

**Comment by tstellar** [comment]
> Basically, folks passing the `-fsized-deallocation` Clang flag (which is off by default in LLVM 18) will get the wrong deallocation size passed to their `operator delete` when storage has been allocated/deallocated by `std::string`. I don't know what the result of passing the wrong size will usually be, I guess it depends on what the custom `operator delete` does, but I could definitely imagine it being a correctness concern.

Is there a reason why gcc has this on by default, but clang doesn't?

**Comment by ldionne** [comment]
> Is there a reason why gcc has this on by default, but clang doesn't?

Basically, we had trouble with rolling it out. It's a wide-reaching change to deploy in the wild because it's effectively an ABI change (call sites that used to call `operator delete(void*)` will now sometimes call `operator delete(void*, size_t)`), and it just took a long time for us to be comfortable and able to make it the default. For example, making `-fsized-deallocation` enabled by default required many rounds of fixing tests and other stuff just within LLVM. Or at least that's my understanding.

**Comment by tstellar** [comment]
OK, if we backport this fix, is there someone willing to do some extra testing on the release branch?  I would like to get some kind of confirmation that the backport fixes the issues and that it doesn't introduce another regression?

**Comment by ldionne** [comment]
@vitalybuka found this with sanitizers (and `-fsized-deallocation`) originally, perhaps he can help with running the tests in that same configuration on the release branch with the cherry-pick?

**Comment by vitalybuka** [comment]
> @vitalybuka found this with sanitizers (and `-fsized-deallocation`) originally, perhaps he can help with running the tests in that same configuration on the release branch with the cherry-pick?

Sure, send me PR, or a branch with a fix.

**Comment by jyknight** [comment]
> There seems to be a claim that Google's using this flag at large scale internally. If it's possible to detail the impact of this regression a bit more, it might help to give at least one "real world" point of view on this.

Google has indeed been using -fsized-deallocation with tcmalloc (which uses the deallocation size information) internally for a long time. However, we also use the libc++ unstable abi, which enables the "alternate" string layout. In that version, the "endian_factor" constant is 1 on little endian machines, rather than 2, so this bug doesn't appear.

**Comment by DimitryAndric** [comment]
I also had to cherry-pick @ldionne's 382f70a877f0 which is a formatting-only change, otherwise d129ea8d2fa3 won't apply cleanly.

**Comment by h-vetinari** [comment]
> Is there a reason why gcc has this on by default, but clang doesn't?

Clang 19 finally enabled it by default (as of https://github.com/llvm/llvm-project/commit/130e93cc26ca9d3ac50ec5a92e3109577ca2e702). IMO this makes it even more important to be able to provide a smooth upgrade path (i.e. people being able to test `-fsized-deallocation` separately before bumping to clang 19), aside from the inherent importance of being C++14 feature (i.e. very old, but which turned out to be exceptionally hard to roll out).

**Review by Dramaga11 — APPROVED** [review]

**Pull request merged**
